src/core/Instrument.cpp

```diff
@@ -79,7 +83,50 @@ bool Instrument::isFromTrack( const Track * _track ) const
 }
 
 
+void Instrument::applyFadeIn(  sampleFrame * buf, const NotePlayHandle * n )
+{
+	// apply only if it's the start of the note
+	if (n->totalFramesPlayed() == 0)
+	{
+		const fpp_t frames = n->framesLeftForCurrentPeriod();
+
+		// zero point crossing counts of all channels
+		int zero_crossings[DEFAULT_CHANNELS] = {0};
+		// maximum zero point crossing of all channels
+		int max_zc = 0;
 
+		// determine the zero point crossing counts
+		for ( fpp_t f = 0; f < frames; ++f )
+		{
+			for ( ch_cnt_t ch=0; ch < DEFAULT_CHANNELS; ++ch )
+			{
+				if ( ( buf[f-1][ch] < 0.0 && buf[f][ch] > 0.0 ) ||
```

Comment:
Not entirely sure that formula is 100% accurate. `abs(1) > abs(1 + -3)` is false AFAIK, so you might miss the occasional crossing this way too?

Reply:
Would adding an extra `|| buf[f][ch] == 0.0` make it catch exact zeros?

Reply:
Yeah you're right, somehow I assumed the first sample could be the one with the biggest absolute value. You could probaly change it to `max(abs(a), abs(b))` to fix it. There's also `abs(a) + abs(b) > abs(a+b)`.

- `max(abs(a), abs(b)) >= abs(a+b) && a != b` should catch everything in 6 comparisons.
- `abs(a) + abs(b) > abs(a+b) || (a != b && (a = 0 || b = 0))` uses the same number but more AND/ORs. It catches non-exact-zero crossings more elegantly IMO, but is a pain to catch the exact ones.
- `((a >= 0 && b <= 0) || (a <= 0 && b>= 0)) && a != b` is five comparisons, if we modify the original solution.

All of these would catch two separate crossings if the sample value goes { -1, 0, 0, 1 }.

How/why is the zero crossing detection actually used in this PR? I haven't figured it out from my brief review.

Reply:
The intuition behind counting zero point crossings is adapting to the frequency of the sound. We need a longer fade in for lower frequencies because the clicking is more audible for lower notes. 

You are absolutely right, current algorithm misses the crossing in [1, 0, -1]. However, the problem with catching [1, 0] and [0, -1] as zero point crossings is that [1, 0, -1] will be count as two zero point crossings. We don't want that because if you plot the amplitude vs time graph of [1, 0, -1], you can see that it crosses the time axis only once. Furthermore, [1, -1, 1] clearly has more frequency than [1, 0, -1], hence these should not have the same zero point crossing counts.

Although using abs function looks shorter, it might be slower (depending on the implementation) if it uses conditional branching. The same goes for max function. Also I don't think that it contributes to the readability of the code.

`(a >= 0 && b < 0) || (a <= 0 && b > 0)` might be a good solution. [1, 0, -1] will be count as one zero point crossing since [1, 0] returns false and [0, -1] returns true.

Reply:
>We don't want [duplicate crossings]

Yeah, that's why I mentioned it.

>using abs function [...]

Aside from the first (faulty) check, I wasn't suggesting any of these. I was just presenting the "fully developed" abs solutions for comparison purposes.

`(a >= 0 && b < 0) || (a <= 0 && b > 0)` looks great.

Reply:
Why not ask the NotePlayHandle for its frequency and calculate a suitable fade in from that?

Reply:
Because the frequency may change with coarse detuning, frequency modulation etc.

Reply:
> `(a >= 0 && b < 0) || (a <= 0 && b > 0)` might be a good solution. [1, 0, -1] will be count as one zero point crossing since [1, 0] returns false and [0, -1] returns true.

Yes, though `[1, -0.000000001, 1]` or `[-1, 0.000000001, -1]` will count as 2.

I'm not sure why we deviate from zyn's original algorithm which counts only positive crossings. Improvements are always cool, but zyn's algorithm sounds absolutely perfect for >10 years and I wonder if we'll get better than absolutely perfect :grin: .

Reply:
Though zyn's algorithm only works until a buffersize of 256. At 32, it's useless. This algorithm will have the same difficulties. There are different approaches, like using the note height (@softrabbit mentioned it, but as discussed, there is modulation and stuff, and there are also harmonics with a different frequency), or, in case of 32 samples, using this algorithm iteratively to get step-wise to the full volume (i.e. multiple runs).

Maybe, instead of trying, we should ask for a good algorithm on [dsp.stackexchange.com](https://dsp.stackexchange.com) , [linux-audio-developper mailing list](https://lists.linuxaudio.org/listinfo/linux-audio-dev/) or [The KVR audio forum](https://www.kvraudio.com/forum/).

Reply:
> [1, -0.000000001, 1] or [-1, 0.000000001, -1] will count as 2.

In that case we can add some margin of error:
`(a >= 0 && b < -0.000000001) || (a <= 0 && b > 0.000000001)`
Perhaps this margin of error should be `std::numeric_limits<float>::epsilon()` or a constant.

> I'm not sure why we deviate from zyn's original algorithm which counts only positive crossings. 

Frankly, I don't understand why it only counts positive crossings (for a little bit of speed, maybe?). So I decided to count both.

> At 32, it's useless. This algorithm will have the same difficulties.

As a solution, I thought about calculating the length in the first iteration and the saving it somewhere. In the next iterations we can just use that length to calculate the proper amplitude. However, there are problems with that. At a buffersize of 32, we can only see a tiny fraction of the wave, so the length calculation may not be accurate. Also, I have no idea where the length should be saved.

Another way to solve this problem might be to calculate the steepness of the wave. If function starts at 0 and gradually increases to 1 like a sine wave, we don't need any fade in. If function starts at 1 and gradually decreases to 0, similarly to the example in the screenshot, we will need to smooth the beginning of the wave, because it actually jumps from 0 to 1 in the beginning. 

I am not an expert on this topic, probably there are many people who are more qualified than me on the sites you have mentioned.

Reply:
I just posted in the [linux audio developer list](http://linux-audio.4202.n7.nabble.com/Click-free-fade-in-algorithm-for-synths-td108758.html) ("LAD"). Let's see if this will help us.

Considering your solutions, the 1st sounds doable. The first 32 samples don't reveal much useful information, but you can still use it and just increase e.g. up to ~ 1/8. If you were much wrong, maybe one can not hear it, because it's just 1/8. And at 2/8, you get more exact, still with a bit error, but it's still not crucial at 2/8. And so on. The more crucial it gets, the more information you have. Anyways, I'd prefer waiting for LAD.

Reply:
The only think applicable I found in the mailing list were the links from Ardour:

* https://github.com/Ardour/ardour/blob/master/libs/ardour/ardour/disk_reader.h#L135
* https://github.com/Ardour/ardour/blob/master/libs/ardour/disk_reader.cc#L1445
* https://github.com/Ardour/ardour/blob/master/libs/ardour/amp.cc#L163

Has anyone time to read them and check what that algorithm actually does?

Reply:
>Has anyone time to read them and check what that algorithm actually does?

`DeclickAmp` has a _g and _a member values. _a is dependent on the sample rate, and it affects how fast _g increases. Every sample is multiplied with _g until it reaches the `target`, which is a parameter of `apply_gain`. When _g reaches the target, it uses the `Amp`'s `apply_simple_gain` method.

 `apply_gain` method of `DeclickAmp` iterates through the buffer in groups of maximum 16. `g`, which is the local version of `_g`, gets updated when the iteration of a group ends. So g is not updated after each sample, it only gets updates once every 16 samples. g is updated using `g += a * (target - g);` where `a==_a`. Thus the increase rate of g gets smaller and smaller. When the difference between `g` and `target` becomes smaller than 1e-5, `g` is set to `target`, ending the fade in phase.

So this algorithm doesn't do anything with the shape/frequency/zero point crossings of the wave. It only depends on the sample rate. And it decays exponentialy, which makes sense because the clicking happens at the beginning.

Some people on the mailing list have objected to our approach, saying that a synth should be controllable, it shouldn't try to outsmart its user. I respect the developers of ZynAddSubFX, but their approach might be obsolete/unwanted. Perhaps all we need is a controllable fade in whose length depends on the frequency of the note, so that the low notes will have longer fade in times. If the frequency of the notes can change with modulation, detuning etc., the user can just adjust the parameters of the fade in accordingly.

Reply:
Then let's let the algorithms compete. You already have one for the zero crossings, one dependent on the sample rate ("fixed length") should be easy to implement.

You could do it like this: Add a counter to fade in, which increases by one for every fade in. If it's even, use zero crossings, if it's odd, use "fixed" length (we could even try some more, I'd be interested to experiment there as well). This will make it easy to experiment and find out which fades are best (without recompiling).

I'm sorry that this sounds like more work then you could have expected from the PR. But these are important experiments that may be reused for many other LMMS synths and even zyn, so this is very valuable research.

Reply:
I'm on it, I will provide the visualizations and output files for each algorithm when it's done. This review has turned into a long conversation, so I will resolve it here and post the results as a new comment.